Reorganized layout of project

codeinred opened the pull request:
Split project into header files and example programs

codeinred committed:
Added support for ccls

ccls is an open-source C++ language server. See: https://github.com/MaskRay/ccls

codeinred committed:
Reorganized project as a library

Header files were moved into the include/ directory; example programs like the cat implementation were moved into examples/ directory. CMakeLists was updated to target cxxeffect as a library.

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,83 @@
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
```
Why was this removed? Isn't it helpful for IDEs and such?

codeinred replied:
We could add it back in, although it'd have to be targeted towards C++20 since we'll be integrating coroutines!

codeinred replied:
CCLS reads the build configuration from compile_commands.json so it doesn't need set_property, although if you know of other IDEs that rely on it, retaining it could be a good idea

s5bug replied:
VSCode's CMake stuff relies on it by default. CLion also relies on it 😄

s5bug commented on examples/cat.cpp:
```diff
@@ -1,6 +1,6 @@
-#include "eff.hpp"
-#include "stream.hpp"
-#include "task.hpp"
+// #include <cxxeffect/eff.hpp>
```
If it's not needed, it should just be removed. Alternatively, perhaps `eff.hpp` should import both `stream` and `task` for you?

codeinred replied:
This was an oversight; I'll remove it

s5bug replied:
What do you think about `eff.hpp` including all of the library headers by default? That would make SFH generation easier too, wouldn't it?

codeinred replied:
> If it's not needed, it should just be removed. Alternatively, perhaps `eff.hpp` should import both `stream` and `task` for you?

I removed this include!

codeinred committed:
Removed include that was commented out

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,83 @@
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
```
Alright, so just adding back the property for CXX20 should be good here. I'm not a CMake expert, would that negate the need for the `-std`/`/std` compiler flags? I feel like it should.

codeinred replied:
Probably! I'll test it!

codeinred replied:
Update: We still need the -std / std compile flags. CMake doesn't add `-std=c++20` even when we have `set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 20)`

s5bug replied:
This seems relevant: https://stackoverflow.com/a/20165519/4580217

Esp. for `-fcoroutines`

s5bug replied:
CMake doesn't have a feature flag for coroutines, and `target_compile_features(cxxeffect PRIVATE cxx_std_20)` should be automatically added when `CXX_STANDARD` is set to 20...

s5bug replied:
Ah, what happens when you set `CXX_STANDARD_REQUIRED` instead of `CXX_STANDARD`?

codeinred replied:
That still doesn't do it, but if we require C++20 as part of the interface of cxxeffect it works! i will update CMakeLists.txt.
```cmake
target_compile_features(cxxeffect INTERFACE cxx_std_20)
```

s5bug replied:
Alright, we should do both `CXX_STANDARD_REQUIRED` and `cxx_std_20` then 👍 

Really odd that this happens... I feel like CMake should be doing the latter from the former.

codeinred committed:
Added set_property CXX_STANDARD

codeinred committed:
Merge branch 'restructure-project' of github.com:s5bug/cxxeffect into restructure-project

codeinred commented:
I added back set_property per your request!

s5bug commented:
Unrelated to the PR content, but to prevent "Merged x into x" commits, you should be `rebase`ing instead of `pull`ing (you can do both at once with `git pull --rebase` on the command line; no clue how to do it in GitKraken).

codeinred commented:
Oh! I appreciate the tip about `git pull --rebase` - I didn't know about it!

codeinred committed:
Targeted cxx_std_20

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,84 @@
 
+target_compile_features(cxxeffect INTERFACE cxx_std_20)
+
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
```
Comment no longer needed here, I think?

codeinred replied:
Yes. I'll add comments explaining the additional compiler options like "/EH" and remove that comment.

codeinred commented:
Added:
```cmake
set_property(TARGET cxxeffect PROPERTY CXX_STANDARD_REQUIRED 20)

target_compile_features(cxxeffect INTERFACE cxx_std_20)
```
And removed the target_compile_options that were made redundant.

If there are no other issues I think we're good to merge!

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,84 @@
+
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
+if (MSVC)
+    target_compile_options(cxxeffect INTERFACE "/EHsc")
```
What does `/EHsc` do? I'd add a comment here to explain to people like me who aren't familiar with MSVC.

codeinred replied:
It makes msvc exception handling standards-compliant. Otherwise, msvc won't do stack unwinding

codeinred replied:
Reviewing https://docs.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-160, we should perhaps change it to "/EHa"

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,84 @@
+# Provides install target, so users can install the library on their system
+# if they like
+
+option(CXXEFFECT_Install "Install CMake targets during install step." ON)
```
Doesn't `cmake -B` have an option to choose the build target? If this is a separate build step, make it a build target if CMake can do that (`--target install`).

Or, is it a flag for i.e. locating a CMake package? Should it then be something like `CXXEFFECT_INSTALL_CMAKE_FILES`? (I have to believe there's some community standard for it, but it seems odd to me to not be in all-caps)

codeinred replied:
I think some of the changes we're making are starting to fall outside of the scope of this pull request. I'm going to push the updated CMake documentation. How about we merge this pull request after that, and then create a separate branch for build system updates?

s5bug replied:
Agreed. Things should follow the naming conventions though, so I believe this should still be `CXXEFFECT_INSTALL`, but I could be entirely wrong.

codeinred replied:
Huh. It turns out that that option is completely unnecessary for installation to work. Even with that line completely removed, you can still install via running
```bash
sudo cmake --install .
```
In the build directory

codeinred replied:
I'm just gonna remove it

codeinred replied:
Update: I removed it since it didn't do anything. Are we ready to merge into main?

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,84 @@
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+    file(GLOB all_tests "examples/*.cpp")
```
Rename to `all_examples`

s5bug commented on CMakeLists.txt:
```diff
@@ -1,17 +1,84 @@
+# If we're building this project, and not something that has this project as a dependency
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```
What do you think about a flag `CXXEFFECT_BUILD_EXAMPLES` that's by-default `ON`, and gets &&ed with this?

codeinred replied:
I've updated all_tests to all_examples, and added documentation for the "/EH" flag. I think some of these changes and improvements should be made on a separate branch specifically for the build system. Do you think that would work?

codeinred committed:
Fixed comments and variable names; updated to /EHa

codeinred committed:
Removed option that didn't do anything

s5bug reviewed (approved):
Thank you so much for your work on this!

codeinred commented:
Thank you so much for being so patient with me!
